### PR TITLE
BTFS-2008 & BTFS-1973

### PR DESCRIPTION
**Pull request by laipogo**
BTFS-1973: Wallet BE: New command to read exchange transaction history

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior?** (You can also refer to a JIRA ticket here)

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)

* **Description of changes**

---

* **Please check if the PR fulfills these requirements**
- [ ] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ ] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ ] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [ ] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [ ] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [ ] Code changes have run through `go fmt`
- [ ] Code changes have run through `go mod tidy`
- [ ] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

File: core/commands/wallet.go [hunk 1]
```diff
@@ -3,14 +3,13 @@ package commands
 import (
 	"errors"
 	"fmt"
+	"github.com/TRON-US/go-btfs/core/commands/cmdenv"
+	"github.com/TRON-US/go-btfs/core/wallet"
+	walletpb "github.com/TRON-US/go-btfs/protos/wallet"
```
Comment (taiyangc) [inline core/commands/wallet.go:8]: organize

File: core/wallet/transaction.go [hunk 2]
```diff
@@ -117,7 +129,7 @@ func DepositRequest(prepareResponse *exPb.PrepareDepositResponse, privateKey *ec
 }
 
 // Continuous call ConfirmDeposit until it responses a FAILED or SUCCESS.
-func ConfirmDepositProcess(prepareResponse *exPb.PrepareDepositResponse, privateKey *ecdsa.PrivateKey) {
+func ConfirmDepositProcess(n *core.IpfsNode, prepareResponse *exPb.PrepareDepositResponse, privateKey *ecdsa.PrivateKey) {
 	ctx := context.Background()
```
Comment (taiyangc) [inline core/wallet/transaction.go:133]: can you double check all wallet operations involve exchange calls should have ctx passed down (req.Context()) from the command, only use context.Background() if it's called from goroutine. I think we missed this in past reviews, check all deposit/withdrawal/confirm commands.

File: core/wallet/transaction.go [hunk 3]
```diff
@@ -50,10 +56,16 @@ func Deposit(ledgerAddr []byte, amount int64, privateKey *ecdsa.PrivateKey, runD
 	}
 	log.Debug(fmt.Sprintf("Call Deposit API success, id: [%d]", prepareResponse.GetId()))
 
+	err = PersistTx(n.Repo.Datastore(), n.Identity.Pretty(), prepareResponse.GetId(),
+		amount, BttWallet, InAppWallet, StatusPending)
+	if err != nil {
+		log.Error(fmt.Sprintf("Record deposit tx failed, reasons: [%v]", err))
```
Comment (taiyangc) [inline core/wallet/transaction.go:62]: return err? (this is the same as "lost" on UI ... better warn user, log.Error does not return err back to UI)

File: core/wallet/transaction.go [hunk 4]
```diff
@@ -117,7 +129,7 @@ func DepositRequest(prepareResponse *exPb.PrepareDepositResponse, privateKey *ec
 }
 
 // Continuous call ConfirmDeposit until it responses a FAILED or SUCCESS.
-func ConfirmDepositProcess(prepareResponse *exPb.PrepareDepositResponse, privateKey *ecdsa.PrivateKey) {
+func ConfirmDepositProcess(n *core.IpfsNode, prepareResponse *exPb.PrepareDepositResponse, privateKey *ecdsa.PrivateKey) {
```
Comment (taiyangc) [inline core/wallet/transaction.go:132]: Not sure why this doesn't return err. It should, for easier err handling.

File: core/wallet/transaction.go [hunk 5]
```diff
@@ -260,11 +287,27 @@ func Withdraw(ledgerAddr, externalAddr []byte, amount int64, privateKey *ecdsa.P
 		log.Error(fmt.Sprintf("Withdraw error, reasons: [%v]", err))
 		return 0, 0, err
 	}
+
+	err = PersistTx(n.Repo.Datastore(), n.Identity.Pretty(), prepareResponse.GetId(), amount,
+		InAppWallet, BttWallet, StatusPending)
+	if err != nil {
+		log.Error(fmt.Sprintf("Persist tx error, reasons: [%v]", err))
```
Comment (taiyangc) [inline core/wallet/transaction.go:294]: same as before, err return

File: core/wallet/transaction.go (hunk 2 above)
Comment (laipogo) [reply core/wallet/transaction.go:133]: okay, will do.

File: core/wallet/transaction.go (hunk 3 above)
Comment (laipogo) [reply core/wallet/transaction.go:62]: my concern is if we return err for PersistTx, client side may do deposit/withdraw again, but actually it's already done.

File: core/commands/wallet.go [hunk 6]
```diff
@@ -161,23 +177,24 @@ var walletBalanceCmd = &cmds.Command{
...
-		return cmds.EmitOnce(res, &MessageOutput{s})
-	},
-	Encoders: cmds.EncoderMap{
-		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *MessageOutput) error {
-			fmt.Fprint(w, out.Message)
-			return nil
-		}),
+		return cmds.EmitOnce(res, &BalanceResponse{
+			BtfsWalletBalance: tronBalance,
+			BttWalletBalance:  lederBalance,
```
Comment (taiyangc) [inline core/commands/wallet.go:189]: typo: ledgerBalance

**Comment by laipogo** [comment]
will reopen after `btfs init` issue fixed.

**Pull request closed without merging**